### PR TITLE
search/jetbrains: introduce type ContentMatchWithLineMatches for clients still using line matches

### PR DESCRIPTION
**Pull request by tbliu98**
Following up on [this feedback](https://github.com/sourcegraph/sourcegraph/pull/42271#issuecomment-1268286970)

The `ContentMatch` type has been migrated to use chunk matches, which is the default result type requested by the web app. However, the JetBrains plugin still relies on the `LineMatch` result type. This PR adds a new type, `ContentMatchWithLineMatches`, which continues to support line matches. The `type` property of this type has value `contentWithLineMatches`. Separating these two content match types allows for a type-safe way for the web app to request chunk matches while the JetBrains plugin continues to request line matches.

## Test plan
Manually test a few content and symbol searches in the plugin locally
CI

### CODE REVIEW

File: client/search-ui/src/results/StreamingSearchResultsList.tsx [hunk 1]
```diff
@@ -203,6 +204,14 @@ export const StreamingSearchResultsList: React.FunctionComponent<
...
                         />
                     )
+                default:
+                    // eslint-disable-next-line no-console
+                    console.error(`Unexpected result type, ${result.type}`)
+                    return <Alert className="d-flex m-3" variant="info">
+                        <Text className="m-0">
+                            <strong>Unexpected result type: </strong> <Code>{result.type}</Code>
+                        </Text>
+                    </Alert>
```
Comment (tbliu98) [inline client/search-ui/src/results/StreamingSearchResultsList.tsx:214]: What is best practice for handling scenarios like this? 

In practice this should never happen in the web app. The value of `result.type` should always be one of the supported values.

Comment (vdavid) [reply client/search-ui/src/results/StreamingSearchResultsList.tsx:214]: I think an exhaustive check would be great here.
Then CI would catch it and we won’t ever need to show the Alert.

Like:
```suggestion
                default:
                    const exhaustiveCheck: never = result.type
                    // eslint-disable-next-line no-console
                    console.error(`Unexpected result type, ${exhaustiveCheck as string}`)
                    return <></>
```

File: client/search-ui/src/results/StreamingSearchResultsList.tsx [hunk 2]
```diff
@@ -30,6 +30,7 @@ import {
 import { SettingsCascadeProps } from '@sourcegraph/shared/src/settings/settings'
 import { TelemetryProps } from '@sourcegraph/shared/src/telemetry/telemetryService'
 import { ThemeProps } from '@sourcegraph/shared/src/theme'
+import {Alert, Code, Text} from '@sourcegraph/wildcard';
```
Comment (vdavid) [inline client/search-ui/src/results/StreamingSearchResultsList.tsx:33]: ...and then we don't need this
```suggestion
```

File: client/search-ui/src/results/StreamingSearchResultsList.tsx [hunk 3]
```diff
@@ -203,6 +204,14 @@ export const StreamingSearchResultsList: React.FunctionComponent<
                             as="li"
                         />
                     )
+                default:
```
Comment (vdavid) [inline client/search-ui/src/results/StreamingSearchResultsList.tsx:207]: I think we also need `case 'contentWithLineMatches'`, probably in line 150 above, or am I missing something?

File: client/jetbrains/webview/src/search/js-to-java-bridge.ts [hunk 4]
```diff
@@ -158,7 +158,7 @@ export async function indicateFinishedLoading(
 export async function onPreviewChange(match: SearchMatch, lineOrSymbolMatchIndex?: number): Promise<void> {
     try {
         const initiationDateTime = new Date()
-        if (match.type === 'content' || match.type === 'path' || match.type === 'symbol') {
+        if (match.type === 'contentWithLineMatches' || match.type === 'path' || match.type === 'symbol') {
```
Comment (vdavid) [inline client/jetbrains/webview/src/search/js-to-java-bridge.ts:161]: And what should happen in the case of `match.type` === 'content'`?
This is a question to all the cases where `match.type === 'content' was changed to `match.type === 'contentWithLineMatches'`.

File: client/search-ui/src/results/StreamingSearchResultsList.tsx (hunk 3 above)
Comment (tbliu98) [reply client/search-ui/src/results/StreamingSearchResultsList.tsx:207]: The web app doesn't support LineMatches anymore, so the case where `result.type === 'contentWithLineMatches'` would be handled by the default case here, where we want to surface that the backend returned an unexpected type

File: client/jetbrains/webview/src/search/js-to-java-bridge.ts (hunk 4 above)
Comment (tbliu98) [reply client/jetbrains/webview/src/search/js-to-java-bridge.ts:161]: `match.type === 'content'` now indicates that the `chunkMatches` request flag was set to true in the search request ([this is now the default in the web app](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/search/results/StreamingSearchResults.tsx?L109&subtree=true)). 

If the `chunkMatches` flag is not set in the request then it defaults to false, the backend will return line matches, and `match.type` will be `contentWithLineMatches`. This is handled by [the changes to `search.go` and `events.go`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/compare/7df36147f15fa4bf83bf99af50addbd8d96dead7...66aaedfaf79b5c48ac3500e02906b6e96fc5481a?visible=12#diff-2ced8430ceda2dc5e243bd61c3d8d987).

The JetBrains plugin does not set `chunkMatches: true` in its streaming search request, so the value of `type` will always be set to `contentWithLineMatches` in the backend for requests originating from this plugin. From what I can tell, I don't think the plugin re-uses any of the search UI components either, so this should preserve existing behavior. I attempted to migrate the plugin to chunk matches as part of the [initial PR](https://github.com/sourcegraph/sourcegraph/pull/42271#issuecomment-1268286970) but it wasn't as simple as I had hoped, so I pivoted to trying to limit the scope of the change.

TLDR; `match.type` should always be `contentWithLineMatches` for search requests originating from the JetBrains plugin

File: client/search-ui/src/results/StreamingSearchResultsList.tsx (hunk 1 above)
Comment (tbliu98) [reply client/search-ui/src/results/StreamingSearchResultsList.tsx]: Thanks for the suggestion! I'm getting this build-ts error: `Type 'string' is not assignable to type 'never'.` TS noob here, is there a way around this to achieve what you're suggesting?

File: client/jetbrains/webview/src/search/js-to-java-bridge.ts (hunk 4 above)
Comment (philipp-spiess) [reply client/jetbrains/webview/src/search/js-to-java-bridge.ts:161]: @tbliu98 `match.type` comes from the backend, right? So this might still be `content` when connecting to older instances? For backward comp, we may need to handle both cases. As you already mentioned, a new server version should return `contentWithLineMatches` so `content` would indicate they're on an older server version?

File: client/search-ui/src/results/StreamingSearchResultsList.tsx (hunk 1 above)
Comment (philipp-spiess) [reply client/search-ui/src/results/StreamingSearchResultsList.tsx]: @tbliu98 Exhaustive types only work for union types. If TypeScript can't enumerate all possible values (and with `string` that is impossible lol) this doesn't work :(

**Comment by tbliu98** [comment]
Closing this in favor of https://github.com/sourcegraph/sourcegraph/pull/42902 which addresses the core concern without any implications for backwards compatibility
